src/Kestrel/KestrelConfigurationBuilder.cs

```diff
@@ -0,0 +1,306 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Https;
+using Microsoft.AspNetCore.Server.Kestrel.Https.Internal;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Server.Kestrel
+{
+    public class KestrelConfigurationBuilder : IKestrelConfigurationBuilder
+    {
+        internal KestrelConfigurationBuilder(KestrelServerOptions options, IConfiguration configuration)
+        {
+            Options = options ?? throw new ArgumentNullException(nameof(options));
+            Configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+        }
+
+        public KestrelServerOptions Options { get; }
+        public IConfiguration Configuration { get; }
+        private IDictionary<string, Action<EndpointConfiguration>> EndpointConfigurations { get; }
+            = new Dictionary<string, Action<EndpointConfiguration>>(0);
+        // Actions that will be delayed until Build so that they aren't applied if the configuration builder is replaced.
+        private IList<Action> EndpointsToAdd { get; } = new List<Action>();
+
+        /// <summary>
+        /// Specifies a configuration Action to run when an endpoint with the given name is loaded from configuration.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="configureOptions"></param>
+        public KestrelConfigurationBuilder Endpoint(string name, Action<EndpointConfiguration> configureOptions)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            EndpointConfigurations[name] = configureOptions ?? throw new ArgumentNullException(nameof(configureOptions));
+            return this;
+        }
+
+        /// <summary>
+        /// Bind to given IP address and port.
+        /// </summary>
+        public KestrelConfigurationBuilder Endpoint(IPAddress address, int port) => Endpoint(address, port, _ => { });
+
+        /// <summary>
+        /// Bind to given IP address and port.
+        /// The callback configures endpoint-specific settings.
```

Comment:
Nit: If we're going to include the sentence about the configure callback, we should do so for all applicable overloads.

Reply:
Removed